File: make.py
#!/usr/bin/env python3

import subprocess
from os import rename as mv, mkdirs as mkd, rmdir, remove as rm

def system(cmd):
    return subprocess.check_output(cmd, shell=True).decode('utf-8')

def flutter(cmd):
    return system('flutter ' + cmd)

def strip(files):
    return system(STRIP + ' ' + files)

def sed(file, old, new):
    f = open(file, 'r+')
    f.write(f.read().replace(old, new))
    f.close()

VERSION = '2.5.1'
ACTUAL_VERSION = VERSION + '.' + system('git rev-parse @ | cut -c 1-7')

FLAGS = '--release --suppress-analytics'
BIN_FLAGS = FLAGS + '--split-debug-info=/tmp --obfuscate'
IOS_FLAGS = BIN_FLAGS
APK_FLAGS = BIN_FLAGS + '--shrink --target-platform android-arm,android-arm64,android-x64'
AAB_FLAGS = APK_FLAGS
WIN_FLAGS = BIN_FLAGS
GTK_FLAGS = BIN_FLAGS
MAC_FLAGS = BIN_FLAGS
WEB_FLAGS = FLAGS + '--csp'

TEST_FLAGS = '--coverage -j 100 --test-randomize-ordering-seed random'

IOS_BUILD_DIR = 'build/ios/Release-iphoneos/Runner.app'
IOS_STRIP_LIST = IOS_BUILD_DIR + '/Runner ' + \
                 IOS_BUILD_DIR + '/Frameworks/App.framework/App ' + \
                 IOS_BUILD_DIR + '/Frameworks/Flutter.framework/Flutter ' + \
                 IOS_BUILD_DIR + '/Frameworks/shared_preferences.framework/shared_preferences '
MAC_BUILD_DIR = 'build/macos/Build/Products/Release/Amplessimus.app'
MAC_STRIP_LIST = MAC_BUILD_DIR + '/Contents/Runner ' + \
                 MAC_BUILD_DIR + '/Contents/Frameworks/App.framework/Versions/A/App ' + \
                 MAC_BUILD_DIR + '/Contents/Frameworks/FlutterMacOS.framework/Versions/A/FlutterMacOS ' + \
                 MAC_BUILD_DIR + '/Contents/Frameworks/shared_preferences_macos.framework/Versions/A/shared_preferences_macos ' + \
                 MAC_BUILD_DIR + '/Contents/Frameworks/*.dylib'
STRIP = 'strip -u -r'
BITCODE_STRIP = 'xcrun bitcode_strip'

OUTPUT_DIR = 'bin'
OUTPUT_APK = OUTPUT_DIR + '/' + ACTUAL_VERSION + '.apk'
OUTPUT_AAB = OUTPUT_DIR + '/' + ACTUAL_VERSION + '.aab'
OUTPUT_DEB = OUTPUT_DIR + '/' + ACTUAL_VERSION + '.deb'
OUTPUT_DMG = OUTPUT_DIR + '/' + ACTUAL_VERSION + '.dmg'
OUTPUT_IPA = OUTPUT_DIR + '/' + ACTUAL_VERSION + '.ipa'
OUTPUT_GTK = OUTPUT_DIR + '/' + ACTUAL_VERSION + '.linux'
OUTPUT_WEB = OUTPUT_DIR + '/' + ACTUAL_VERSION + '.web'
OUTPUT_WIN = OUTPUT_DIR + '/' + ACTUAL_VERSION + '.win'

TMP = 'tmp'
IPA_DIR = 'Payload'
TMP_IPA_DIR = TMP + '/' + IPA_DIR
TMP_DEB_DIR = TMP + '/deb'
TMP_DMG_DIR = TMP + '/dmg'
TMP_DMG = TMP + '/tmp.dmg'


def ci():
    mkdirs()
    replaceversions()
    iosapp()
    ipa()
    apk()
    cleanartifacts()
    rollbackversions()

def android():
    mkdirs()
    replaceversions()
    apk()
    aab()
    cleanartifacts()
    rollbackversions()

def ios():
    mkdirs()
    replaceversions()
    iosapp()
    ipa()
    deb()
    cleanartifacts()
    rollbackversions()

def web():
    mkdirs()
    replaceversions()
    webbuild()
    cleanartifacts()
    rollbackversions()

def win():
    mkdirs()
    replaceversions()
    winx64()
    cleanartifacts()
    rollbackversions()

def mac():
    mkdirs()
    replaceversions()
    macapp()
    macdmg()
    cleanartifacts()
    rollbackversions()

def linux():
    mkdirs()
    replaceversions()
    linux64()
    cleanartifacts()
    rollbackversions()

def replaceversions():
    sed('pubspec.yaml', '0.0.0-1', VERSION)
    sed('lib/values.dart', '0.0.0-1', VERSION)

# TODO: call this always
def rollbackversions():
	mv('pubspec.yaml.def', 'pubspec.yaml')
	mv('lib/values.dart.def', 'lib/values.dart')

def iosapp():
	system('flutter build ios ' + IOS_FLAGS)
	system(BITCODE_STRIP + ' ' + IOS_BUILD_DIR + '/Frameworks/Flutter.framework/Flutter -r -o tmpfltr')
	mv('tmpfltr', IOS_BUILD_DIR + '/Frameworks/Flutter.framework/Flutter')
	system('rm -f ' + IOS_BUILD_DIR + '/Frameworks/libswift*')
	system(STRIP + ' ' + IOS_STRIP_LIST)

def ipa():
	cp -rp $(IOS_BUILD_DIR) $(TMP_IPA_DIR)
	rmdir(OUTPUT_IPA)
	cd $(TMP) && zip -r -9 ../$(OUTPUT_IPA) $(IPA_DIR)

# http://www.saurik.com/id/7
# but its broken...
def deb():
	cp -rp $(IOS_BUILD_DIR) $(TMP_DEB_DIR)/Applications/
	VERSION=$(VERSION) BUILD_DIR=$(IOS_BUILD_DIR) INPUT=control.def OUTPUT=$(TMP_DEB_DIR)/DEBIAN/control sh sedit.sh
	COPYFILE_DISABLE= COPY_EXTENDED_ATTRIBUTES_DISABLE= dpkg-deb -Sextreme -z9 --build $(TMP_DEB_DIR) $(OUTPUT_DEB)

def cydiainfo():
	VERSION=$(VERSION) BUILD_DIR=$(IOS_BUILD_DIR) DEB=$(OUTPUT_DEB) INPUT=Packages.def OUTPUT=$(OUTPUT_DIR)/Packages sh sedit.sh
	gzip -9 -c $(OUTPUT_DIR)/Packages > $(OUTPUT_DIR)/Packages.gz

def apk():
	flutter build apk $(APK_FLAGS)
	mv build/app/outputs/apk/release/app-release.apk $(OUTPUT_APK)

def aab():
	flutter build appbundle $(AAB_FLAGS)
	mv build/app/outputs/bundle/release/app-release.aab $(OUTPUT_AAB)

def webbuild():
	flutter channel master
	flutter upgrade
	flutter config --enable-web
	flutter build web $(WEB_FLAGS)
	mv build/web $(OUTPUT_WEB)

def winx64():
	flutter channel master
	flutter upgrade
	flutter config --enable-windows-desktop
	flutter build windows $(WIN_FLAGS)
	mv build/windows/x64/Release/Runner $(OUTPUT_WIN)

def linux64():
	flutter channel master
	flutter upgrade
	flutter config --enable-linux-desktop
	flutter build linux $(GTK_FLAGS)
	mv build/linux/release/bundle $(OUTPUT_GTK)

def macapp():
	flutter('channel master')
	flutter('upgrade')
	flutter('config --enable-macos-desktop')
	flutter('build macos ' + MAC_FLAGS)
	$(STRIP) $(MAC_STRIP_LIST) || true

def macdmg():
	cp -rf $(MAC_BUILD_DIR) $(TMP_DMG_DIR)
	ln -s /Applications $(MAC_BUILD_DIR)/Applications
	hdiutil create $(TMP_DMG) -ov -srcfolder $(TMP_DMG_DIR) -fs APFS -volname "Install Amplessimus"
	hdiutil convert $(TMP_DMG) -ov -format UDBZ -o $(OUTPUT_DMG)

def mkdirs():
    mkd(OUTPUT_DIR)
    mkd(TMP_IPA_DIR)
    mkd(TMP_DEB_DIR + '/DEBIAN')
    mkd(TMP_DEB_DIR + '/Applications')
    mkd(TMP_DMG_DIR)

def cleanartifacts():
	rmdir(TMP)

def test():
	flutter('test ' + TEST_FLAGS + ' || /usr/bin/env python3 make.py test')
	system('genhtml -o coverage/html coverage/lcov.info')
	system('lcov -l coverage/lcov.info')

File: make.dart
import 'dart:io';

final shortVersion = '3.7';

late String version;
late String buildNumber;

String get flags => '--release --suppress-analytics ';
String get binFlags => '$flags --build-name=$version '
    '--build-number $buildNumber';
String get iosFlags => binFlags;
//--target-platform android-arm,android-arm64,android-x64
String get apkFlags => '$binFlags --shrink';
String get aabFlags => apkFlags;
String get winFlags => binFlags;
String get linuxX86Flags => '$flags --target-platform linux-x64';
String get linuxARMFlags => '$flags --target-platform linux-arm64';
String get macFlags => binFlags;

final testFlags =
    '--coverage -j 100 --test-randomize-ordering-seed random -r expanded';

Future<String> system(
  String cmd, {
  bool throwOnFail = false,
  bool printInput = true,
  bool printOutput = true,
}) async {
  if (printInput) stderr.writeln(cmd);
  final p = Platform.isWindows
      ? await Process.run('cmd', ['/c', cmd])
      : await Process.run('sh', ['-c', cmd]);
  if (printOutput) {
    stderr.write(p.stderr);
    stderr.write(p.stdout);
  }
  if (p.exitCode != 0 && throwOnFail) throw p.exitCode;
  return p.stdout.trimRight();
}

Future<String> readfile(path) async => File(path).readAsString();
Future writefile(path, content) async => File(path).writeAsString(content);

Future<void> rm(f) async {
  if (await File(f).exists()) {
    await File(f).delete();
  }
}

Future<void> rmd(d) async {
  if (await Directory(d).exists()) {
    await Directory(d).delete(recursive: true);
  }
}

Future mv(from, to) => File(from).rename(to);
Future mvd(from, to) => Directory(from).rename(to);

Future mkdirs(d) => Directory(d).create(recursive: true);

Future<String> md5(String path) =>
    system("md5sum '$path' | cut -d' ' -f1", printOutput: false);

Future<void> flutter(String cmd, {bool throwOnFail = true}) =>
    system('flutter $cmd', throwOnFail: throwOnFail);
Future build(String cmd, String flags) => flutter('build $cmd $flags');

Future<void> strip(String files) => system(
      'strip -u -r $files',
      printOutput: false,
    );

Future unsign(String app) => system('codesign --remove-signature \'$app\'');

Future<void> iosapp() async {
  const buildDir = 'build/ios/Release-iphoneos/Runner.app';
  await build('ios', iosFlags);
  await unsign(buildDir);
  await system(
      'xcrun bitcode_strip $buildDir/Frameworks/Flutter.framework/Flutter -r -o tmpfltr');
  await mv('tmpfltr', '$buildDir/Frameworks/Flutter.framework/Flutter');
  await system('rm -f $buildDir/Frameworks/libswift*');
  await strip('$buildDir/Runner $buildDir/Frameworks/*.framework/*');
}

Future<void> ipa() async {
  //await flutter('build ipa $iosFlags');
  await system('cp -rp build/ios/Release-iphoneos/Runner.app tmp/Payload');
  await rm('bin/$version.ipa');
  await system('cd tmp && zip -r -9 ../bin/$version.ipa Payload');
}

Future<void> apk() async {
  await build('apk', apkFlags);
  await mv('build/app/outputs/flutter-apk/app-release.apk', 'bin/$version.apk');
}

Future<void> aab() async {
  await build('appbundle', aabFlags);
  await mv(
      'build/app/outputs/bundle/release/app-release.aab', 'bin/$version.aab');
}

Future<void> test() async {
  await flutter('test $testFlags');
  await system('genhtml -o coverage/html coverage/lcov.info');
  await system('lcov -l coverage/lcov.info');
}

Future<void> ios() async {
  await iosapp();
  await ipa();
}

Future<void> android() async {
  await apk();
  await aab();
}

Future<void> win() async {
  await flutter('config --enable-windows-desktop');
  await build('windows', winFlags);
  await mvd('build/windows/runner/Release', 'bin/$version.win');
}

Future<void> mac() async {
  await flutter('config --enable-macos-desktop');
  await build('macos', macFlags);
  const bld = 'build/macos/Build/Products/Release/Amplessimus.app';
  await unsign(bld);
  const contents = '$bld/Contents';
  const frameworks = '$contents/Frameworks';
  await system('rm -f $frameworks/libswift*');

  await system('cp -rf $bld tmp/dmg');
  await system('ln -s /Applications "tmp/dmg/Applications"');
  await system('hdiutil create bin/$version.dmg -ov '
      '-srcfolder tmp/dmg -volname "Amplessimus $shortVersion" '
      // 106M UDRW
      // 106M UFBI
      //  86M UDRO
      //  39M UDCO
      //  34M UDZO
      //  31M ULFO
      //  29M UDBZ
      //  25M ULMO
      '-fs APFS -format ULMO');
}

Future<void> linux() async {
  await flutter('config --enable-linux-desktop');
  await build('linux', linuxX86Flags);
  await mvd('build/linux/release/bundle', 'bin/$version-linux-x86_64');
  await cleanup();
  await build('linux', linuxARMFlags);
  await mvd('build/linux/release/bundle', 'bin/$version-linux-arm64');
}

Future<void> ver() async {
  print(version);
}

Future<void> clean() async {
  await rmd('tmp');
  await rmd('build');
  await rmd('bin');
}

Future<void> init() async {
  buildNumber = await system(
    'git rev-list @ --count',
    printInput: false,
    printOutput: false,
    throwOnFail: true,
  );
  version = '$shortVersion.${int.parse(buildNumber) - 1310}';
  await mkdirs('bin');
  await mkdirs('tmp/Payload');
  await mkdirs('tmp/deb/DEBIAN');
  await mkdirs('tmp/deb/Applications');
  await mkdirs('tmp/dmg');
}

Future<void> cleanup() async {
  await rmd('tmp');
  await rmd('build');
}

const targets = {
  'ios': ios,
  'android': android,
  'test': test,
  'win': win,
  'mac': mac,
  'linux': linux,
  'ver': ver,
  'clean': clean,
};

Future<void> main(List<String> argv) async {
  try {
    await init();
    for (final target in argv) {
      if (!targets.containsKey(target)) throw 'Target $target doesn\'t exist.';
      await targets[target]!();
    }
  } catch (e) {
    stderr.writeln(e);
    if (e is Error) stderr.writeln(e.stackTrace);
  }
  await cleanup();
}
